sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/PropertiesAccessCorrectField.cs

```diff
@@ -0,0 +1,256 @@
+﻿using System;
+
+namespace Tests.Diagnostics
+{
+    class NonCompliantClass_FromRspec
+    {
+        private int x;
+        private int y;
+
+        public int X
+        {
+            get { return x; }
+            set { x = value; }
+        }
+
+        public int Y
+        {
+            get { return x; }  // Noncompliant {{Refactor this getter so that it actually refers to the field 'y'}}
+//                       ^
+            set { x = value; } // Noncompliant {{Refactor this setter so that it actually refers to the field 'y'}}
+//                ^
+        }
+    }
+
+    class NonCompliantStruct_FromRspec
+    {
+        private int x;
+        private int y;
+
+        public int X
+        {
+            get { return x; }
+            set { x = value; }
+        }
+
+        public int Y
+        {
+            get { return x; }  // Noncompliant: field 'y' is not used in the return value
+            set { x = value; } // Noncompliant: field 'y' is not updated
+        }
+    }
+
+
+    class NonCompliant_UnderscoresInNamesAndCasing
+    {
+        private int yyy;
+        private int __x__X; // test that underscores and casing in names are ignored
+        public int XX
+        {
+            get { return yyy; }   // Noncompliant {{Refactor this getter so that it actually refers to the field '__x__X'}}
+//                       ^^^
+            set { yyy = value; }  // Noncompliant {{Refactor this setter so that it actually refers to the field '__x__X'}}
+//                ^^^
+        }
+
+        public string _Y___Y_Y_
+        {
+            get { return __x__X; } // Noncompliant
+//                       ^^^^^^
+        }
+    }
+
+    class NonCompliant_FieldTypeIsIgnored
+    {
+        private int aaa;
+        private string aString;
+
+        public string AAA
+        {
+            get { return aString; } // Noncompliant - field called 'aaa' exists, even though type is different
+//                       ^^^^^^^
+            set { aString = value; } // Noncompliant
+//                ^^^^^^^
+        }
+    }
+
+    class NonCompliant_AssigningToExpression
+    {
+        private int aaa;
+        private string aString;
+
+        public string AAA
+        {
+            set { aString = "foo" + value; } // Noncompliant
+//                ^^^^^^^
+        }
+    }
+
+    partial class NonCompliant_PartialClass
+    {
+        private object myProperty;
+    }
+    partial class NonCompliant_PartialClass
+    {
+        private object anotherObject;
+    }
+    partial class NonCompliant_PartialClass
```

Comment:
I originally wrote the rule to register a syntax node action for class and struct declarations. I thought I was done, then I added a test case for _partial_ classes which failed because the issues were being reported for each partial declaration. 
So I changed the rule to use a symbol action it only fires once per definition.